### PR TITLE
boards: stm32: Add Support for Nucleo_G431KB

### PR DESCRIPTION
**Pull request by mariucker**
Add support for Nucleo G431KB. The Nucleo G431KB board features an ARM Cortex-M4 based STM32G431KB MCU
with a wide range of connectivity support and configurations.
Signed-off-by: Mariana Ucker <mariana@freedom.ind.br>

### CODE REVIEW

File: boards/arm/nucleo_g431kb/nucleo_g431kb_defconfig [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+CONFIG_ARM_MPU=y
+
+# Enable HW stack protection
+CONFIG_HW_STACK_PROTECTION=y
+
+# enable pin controller
+CONFIG_PINCTRL=y
+
+
+CONFIG_CAN_STM32FD_CLOCK_SOURCE_PCLK1=y
```
Comment (erwango) [inline boards/arm/nucleo_g431kb/nucleo_g431kb_defconfig:30]: Seems like a deprecated symbol.

File: boards/arm/nucleo_g431kb/nucleo_g431kb.dts [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	status = "okay";
+};
+
+&can1 {
+	pinctrl-0 = <&fdcan1_rx_pa11 &fdcan1_tx_pa12>;
+	pinctrl-names = "default";
+	status = "okay";
+	bus-speed-data = <1000000>;
+	bus-speed = <125000>;
+};
```
Comment (erwango) [inline boards/arm/nucleo_g431kb/nucleo_g431kb.dts:84]: You might want to configure clock here:
```suggestion
	clocks = <&rcc STM32_CLOCK_BUS_APB1 0x02000000>,
		 <&rcc STM32_SRC_FOO FDCAN_SEL(X)>;
};
```

File: boards/arm/nucleo_g431kb/nucleo_g431kb.dts [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+&clk_hsi {
+	status = "okay";
+};
+
+&pll {
+	div-m = <4>;
+	mul-n = <85>;
+      div-p = <7>;
```
Comment (erwango) [inline boards/arm/nucleo_g431kb/nucleo_g431kb.dts:43]: indentation

File: boards/arm/nucleo_g431kb/nucleo_g431kb.dts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	div-q = <2>;
```
Comment (erwango) [inline boards/arm/nucleo_g431kb/nucleo_g431kb.dts:44]: Note that `div-p` and `div-q` are not mandatory if not required.

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for this contribution.
Some comments and then it should be fine.

You'll also need to add a commit message for compliance with contribution guidelines.

For more info about contribution guidelines, I'd advise you can check:
- [Contribution guidelines on docs.zephyrproject.org](https://docs.zephyrproject.org/latest/contribute/guidelines.html)
- [Great presentation dealing with "the best way to get your patches merged"](https://youtu.be/mnVNLM7ynMQ) and the related [Slides](https://static.sched.com/hosted_files/zephyr2022/a3/HOWTO_Get_Your_Zephyr_Patches_Merged_ZDS_2022.pdf)
- Coding Guidelines bot results reported below.

File: boards/arm/nucleo_g431kb/nucleo_g431kb_defconfig (hunk 1 above)
Comment (mariucker) [reply boards/arm/nucleo_g431kb/nucleo_g431kb_defconfig]: It was excluded.

File: boards/arm/nucleo_g431kb/nucleo_g431kb.dts (hunk 4 above)
Comment (mariucker) [reply boards/arm/nucleo_g431kb/nucleo_g431kb.dts]: div-p e div-q were deleted.

File: boards/arm/nucleo_g431kb/nucleo_g431kb.dts (hunk 3 above)
Comment (mariucker) [reply boards/arm/nucleo_g431kb/nucleo_g431kb.dts]: Corrected.

File: boards/arm/nucleo_g431kb/nucleo_g431kb.dts (hunk 2 above)
Comment (mariucker) [reply boards/arm/nucleo_g431kb/nucleo_g431kb.dts:84]: Thanks for the suggestion. It's already been added to the code.

File: boards/arm/nucleo_g431kb/nucleo_g431kb.dts [hunk 5]
```diff
@@ -81,6 +79,8 @@
 	status = "okay";
 	bus-speed-data = <1000000>;
 	bus-speed = <125000>;
+	clocks = <&rcc STM32_CLOCK_BUS_APB1 0x02000000>,
+	<&rcc STM32_SRC_HSI FDCAN_SEL(2)>;
```
Comment (erwango) [inline boards/arm/nucleo_g431kb/nucleo_g431kb.dts:83]: Please fix indentation.

**Comment by erwango** [comment]
Please squash you commit, remove the merge commit and add proper commit message.
This is all mentioned in the doc I provided, please read them carefully.

**Review by FRASTM — DISMISSED** [review]
LGTM, 

1. rmk: why not setting the HSE as clock source of the PLL, there is a 24MHz crystal on the board.

**Comment by erwango** [comment]
@mariucker When fixing the last doc build issue, please also reformat the commit title to the usual format :
`boards: stm32: Add support ....`
Thanks

**Review by FRASTM — DISMISSED** [review]

**Review by erwango — DISMISSED** [review]

**Comment by erwango** [comment]
@mariucker Please fix compliance issue reported:

```
Error: See https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style for more details.

-:349: ERROR:TRAILING_WHITESPACE: trailing whitespace
#349: FILE: boards/arm/nucleo_g431kb/nucleo_g431kb.dts:[22](https://github.com/zephyrproject-rtos/zephyr/actions/runs/3204601971/jobs/5437717998#step:9:23):
+^I$

- total: 1 errors, 0 warnings, 472 lines checked
```

**Review by FRASTM — DISMISSED** [review]

**Review by otavio — DISMISSED** [review]

**Review by gautierg-st — CHANGES_REQUESTED** [review]
Please add the LSI to the dts file and use it as the RTC clock source to conform to the evolution introduced in PR #50104.

File: boards/arm/nucleo_g431kb/nucleo_g431kb.dts [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+	pwm2: pwm {
+		status = "okay";
+		pinctrl-0 = <&tim2_ch1_pa15>;
+		pinctrl-names = "default";
+	};
+};
+
+&rtc {
+	status = "okay";
```
[the remark this answers is not in the document]
Comment (mariucker) [reply boards/arm/nucleo_g431kb/nucleo_g431kb.dts:107]: Corrected.

File: boards/arm/nucleo_g431kb/nucleo_g431kb.dts [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+			gpios = <&gpiob 8 GPIO_ACTIVE_HIGH>;
+			label = "User LD2";
+		};
+	};
+	aliases {
+		led0 = &green_led;
+	};
+
+};
+
```
[the remark this answers is not in the document]
Comment (mariucker) [reply boards/arm/nucleo_g431kb/nucleo_g431kb.dts:35]: Added.

**Review by FRASTM — APPROVED** [review]

**Review by otavio — APPROVED** [review]

**Review by gautierg-st — APPROVED** [review]

File: boards/arm/nucleo_g431kb/doc/index.rst [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+| I2C       | on-chip    | i2c                                 |
++-----------+------------+-------------------------------------+
+| PWM       | on-chip    | pwm                                 |
++-----------+------------+-------------------------------------+
+| DAC       | on-chip    | dac                                 |
++-----------+------------+-------------------------------------+
+| COUNTER   | on-chip    | rtc                                 |
++-----------+------------+-------------------------------------+
+| SPI       | on-chip    | spi                                 |
++-----------+------------+-------------------------------------+
```
Comment (nandojve) [inline boards/arm/nucleo_g431kb/doc/index.rst:120]: Hi @mariucker ,

There is a new table model. I would recommend you to update to that one. See [stm32l1_disco](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/stm32l1_disco/doc/index.rst) as example.

**Review by nandojve — CHANGES_REQUESTED** [review]

File: boards/arm/nucleo_g431kb/doc/index.rst (hunk 8 above)
Comment (erwango) [reply boards/arm/nucleo_g431kb/doc/index.rst:120]: There are > 200 boards using this kind of table. Doesn't feel fair to request a change on this.

**Review by erwango — APPROVED** [review]

**Review by nandojve — APPROVED** [review]

**Comment by kartben** [comment]
@erwango I believe @mariucker will have to force-push an amended commit to re-trigger CI -- @stephanosio can probably confirm if something else could do the trick.

**Comment by stephanosio** [comment]
> @erwango I believe @mariucker will have to force-push an amended commit to re-trigger CI -- @stephanosio can probably confirm if something else could do the trick.

It looks like GitHub was having problems at the time this PR was last pushed. Unfortunately, force-pushing is the only to retrigger CI when this happens.

**Comment by erwango** [comment]
@mariucker Please fix warning reported by CI:
```
/__w/zephyr/zephyr/drivers/timer/stm32_lptim_timer.c:32:2: error: #warning Kconfig for LPTIM source clock (LSI/LSE) is deprecated, use device tree. [-Werror=cpp]
   32 | #warning Kconfig for LPTIM source clock (LSI/LSE) is deprecated, use device tree.
      |  ^~~~~~~
```

**Pull request closed without merging**